### PR TITLE
Introduce typescript

### PR DESCRIPTION
**Pull request by arafat-al-mahmud**

### CODE REVIEW

**Review by zhfahan77 — COMMENTED** [review]
on package.json, in scripts tag

- `start:build` can be `start:watch`
- `start:run` can be `start:build`

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,5 @@ node_modules
 .env
 .gitconfig
 .DS_Store
-.dockerignore
\ No newline at end of file
+.dockerignore
+build/**/*
```
Comment (zhfahan77) [inline .gitignore]: we can just add

`build`

let us ignore the whole dir

File: server/appconfig.ts [hunk 2]
```diff
@@ -1,18 +1,18 @@
-const express = require("express"),
-bodyParser = require("body-parser"),
+import { Request, Response, NextFunction } from 'express';
```
Comment (zhfahan77) [inline server/appconfig.ts]: `import { Request, Response, NextFunction, Application} from 'express';`

File: server/appconfig.ts [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
+const bodyParser = require("body-parser"),
 errorMessages = require("./api/utils/errorMessages");
 
-module.exports.initialize = function(app) {
+module.exports.initialize = function(app: any) {
```
Comment (zhfahan77) [inline server/appconfig.ts]: `module.exports.initialize = function(app: Application) {`

File: server/appconfig.ts [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
 
-    function appUse(app, ...arg) {
+    function appUse(app: any, ...arg: any[]) {
```
Comment (zhfahan77) [inline server/appconfig.ts]: function appUse(app: Application, ...arg: any[]) {

**Review by zhfahan77 — CHANGES_REQUESTED** [review]
i have added my observations 👍

File: server/appconfig.ts (hunk 4 above)
Comment (uraniumreza) [inline server/appconfig.ts]: We shouldn't use `any` type here! Let's import `Application` from express and cast the `app: Application`!

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,66 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    // "lib": [],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
+    // "checkJs": true,                       /* Report errors in .js files. */
```
Comment (uraniumreza) [inline tsconfig.json:9]: `checkJs` is the best feature; don't you think we need to uncomment that? 😄

File: package.json [hunk 6]
```diff
@@ -4,6 +4,9 @@
   "description": "Argus",
   "main": "app.js",
   "scripts": {
+    "start:build": "tsc -w",
+    "start:run": "nodemon build/app.js",
+    "start": "concurrently npm:start:*",
     "test": "mocha **test/*.js --exit"
```
Comment (uraniumreza) [inline package.json:10]: Are we using `mocha` for testing? Didn't see any **dev-dependency** though! I'll vote for `Jest` for our testing library; let's discuss about this @zhfahan77 @arafat-al-mahmud

**Review by uraniumreza — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 5 above)
Comment (arafat-al-mahmud) [reply tsconfig.json:9]: The default value is `true` :D

**Pull request merged**